Userland/Libraries/LibDisassembly/riscv64/IM.cpp

```diff
@@ -0,0 +1,268 @@
+}
+
+NonnullOwnPtr<JumpAndLinkRegister> parse_jalr(u32 instruction)
+{
+    auto raw_parts = RawIType::parse(instruction);
+    return adopt_own(*new (nothrow) JumpAndLinkRegister(raw_parts.imm, raw_parts.rs1, raw_parts.rd));
+}
+
+NonnullOwnPtr<AddUpperImmediateToProgramCounter> parse_auipc(u32 instruction)
+{
+    auto raw_parts = RawUType::parse(instruction);
+    return adopt_own(*new (nothrow) AddUpperImmediateToProgramCounter(raw_parts.imm, raw_parts.rd));
+}
+
+NonnullOwnPtr<ArithmeticImmediateInstruction> parse_op_imm(u32 instruction)
+{
+    auto raw_parts = RawIType::parse(instruction);
+
+    auto operation = ArithmeticImmediateInstruction::Operation::Add;
+    switch (raw_parts.funct3) {
+    case 0b000:
+        operation = ArithmeticImmediateInstruction::Operation::Add;
+        break;
+    case 0b010:
+        operation = ArithmeticImmediateInstruction::Operation::SetLessThan;
+        break;
+    case 0b011:
+        operation = ArithmeticImmediateInstruction::Operation::SetLessThanUnsigned;
+        break;
+    case 0b100:
+        operation = ArithmeticImmediateInstruction::Operation::Xor;
+        break;
+    case 0b110:
+        operation = ArithmeticImmediateInstruction::Operation::Or;
+        break;
+    case 0b111:
+        operation = ArithmeticImmediateInstruction::Operation::And;
+        break;
+    case 0b001:
+        operation = ArithmeticImmediateInstruction::Operation::ShiftLeftLogical;
+        break;
+    case 0b101:
+        if ((instruction & (1 << 30)) == 0)
+            operation = ArithmeticImmediateInstruction::Operation::ShiftRightLogical;
+        else
+            operation = ArithmeticImmediateInstruction::Operation::ShiftRightArithmetic;
+        // Delete the possibly set 10th bit that distinguishes SRLI and SRAI.
+        raw_parts.imm &= ~(1 << 10);
+        break;
+    default:
+        VERIFY_NOT_REACHED();
+    }
+    return adopt_own(*new (nothrow) ArithmeticImmediateInstruction(operation, raw_parts.imm, raw_parts.rs1, raw_parts.rd));
+}
+
+NonnullOwnPtr<ArithmeticImmediateInstruction> parse_op_imm_32(u32 instruction)
+{
+    auto raw_parts = RawIType::parse(instruction);
+
+    auto operation = ArithmeticImmediateInstruction::Operation::Add;
+    switch (raw_parts.funct3) {
+    case 0b000:
+        operation = ArithmeticImmediateInstruction::Operation::AddWord;
+        break;
+    case 0b001:
+        operation = ArithmeticImmediateInstruction::Operation::ShiftLeftLogicalWord;
+        break;
+    case 0b101:
+        if ((instruction & (1 << 30)) == 0)
+            operation = ArithmeticImmediateInstruction::Operation::ShiftRightLogicalWord;
+        else
+            operation = ArithmeticImmediateInstruction::Operation::ShiftRightArithmeticWord;
+        // Delete the possibly set 10th bit that distinguishes SRLIW and SRAIW.
+        raw_parts.imm &= ~(1 << 10);
+        break;
+    default:
+        VERIFY_NOT_REACHED();
```

Comment:
It might be nicer to not crash on invalid instructions in the future, but `VERIFY_NOT_REACHED` is fine for now.